File: java/extract-text/TokenAuth.java
import java.io.File;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.Files;

public class TokenAuth {
        public static void main(String args[]) {
                // token configuration
                String clientId = "[[client_id]]";
                String clientSecret = "[[client_secret]]";
                String scopes = "[[scopes]]";

                // get the access_token from the token endpoint
                String tokenPayload = "{ \"client_id\": \"" + clientId + "\", \"client_secret\": \"" + clientSecret
                                + "\", \"grant_type\": \"client_credentials\", \"scope\": \"" + scopes + "\"  }";
                String token = "";

                try {
                        HttpClient client = HttpClient.newHttpClient();
                        HttpRequest request = HttpRequest.newBuilder()
                                        .uri(URI.create("https://api.intelligent-api.com/v1/token"))
                                        .header("Content-Type", "application/json")
                                        .POST(HttpRequest.BodyPublishers.ofString(
                                                        tokenPayload))
                                        .build();

                        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());

                        String searchKey = "access_token";
                        String tokenResponse = response.body();
                        int keyIndex = tokenResponse.indexOf(searchKey);
                        int valueStart = tokenResponse.indexOf("\"", keyIndex + searchKey.length()) + 3;
                        int valueEnd = tokenResponse.indexOf("\"", valueStart);
                        token = tokenResponse.substring(valueStart, valueEnd);
                } catch (Exception e) {
                        e.printStackTrace();
                        return;
                }

                // use the access_token as a bearer token
                String authorization = "Bearer ".concat(token);

                // populate endpoint parameters
                String fullPathToFile = "[[filepath]]";
                String userAgent = "Intelligent API Sample Java Code";

                // invoke the API endpoint
                try {
                        File file = new File(fullPathToFile);
                        byte[] fileBytes = Files.readAllBytes(file.toPath());
                        HttpClient client = HttpClient.newHttpClient();
                        HttpRequest request = HttpRequest.newBuilder()
                                        .uri(URI.create("https://api.intelligent-api.com/v1/document/text"))
                                        .header("Authorization",
                                                        authorization)
                                        .header("User-Agent",
                                                        userAgent)
                                        .header("Content-Type", "application/octet-stream")
                                        .POST(HttpRequest.BodyPublishers.ofByteArray(
                                                        fileBytes))
                                        .build();

                        HttpResponse<String> response = client.send(request,
                                        HttpResponse.BodyHandlers.ofString());

                        System.out.println("Response Code: " + response.statusCode());
                        System.out.println("Response Body: " + response.body());
                } catch (Exception e) {
                        e.printStackTrace();
                }
        }
}

File: nodejs/extract-expense/token-auth.js
const https = require("https");

// token configuration
const clientId = "[[client_id]]";
const clientSecret = "[[client_secret]]";
const scopes = "[[scopes]]";

let token = "";

async function makeTokenRequest() {
  return new Promise((resolve, reject) => {
    const tokenPostData = JSON.stringify({
      client_id: clientId,
      client_secret: clientSecret,
      grant_type: "client_credentials",
      scope: scopes,
    });

    const tokenOptions = {
      hostname: "api.intelligent-api.com",
      port: 443,
      path: "/v1/token",
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Content-Length": Buffer.byteLength(tokenPostData),
      },
    };

    const tokenRequest = https.request(tokenOptions, (response) => {
      let responseBody = "";
      response.on("data", (chunk) => {
        responseBody += chunk;
      });

      response.on("end", () => {
        try {
          const jsonResponse = JSON.parse(responseBody);
          resolve(jsonResponse.access_token);
        } catch (error) {
          reject(`Error parsing JSON: ${error.message}`);
        }
      });
    });

    tokenRequest.on("error", (e) => {
      reject(`Problem with request: ${e.message}`);
    });

    tokenRequest.write(tokenPostData);
    tokenRequest.end();
  });
}

(async () => {
  try {
    token = await makeTokenRequest();

    // use the access_token as a bearer token
    const authorization = `Bearer ${token}`;

    // populate endpoint parameters
    const fs = require("fs");
    const fullPathToFile = "[[filepath]]";
    const userAgent = "Intelligent API Sample Nodejs Code";

    const fileStream = fs.createReadStream(fullPathToFile);

    const options = {
      hostname: "api.intelligent-api.com",
      port: 443,
      path: "/v1/document/expenses",
      method: "POST",
      headers: {
        Authorization: authorization,
        "Content-Type": "application/octet-stream",
        "User-Agent": userAgent,
      },
    };

    const request = https.request(options, (response) => {
      console.log(`Status Code: ${response.statusCode}`);

      response.on("data", (chunk) => {
        console.log("Response Body:", chunk.toString());
      });
    });

    request.on("error", (e) => {
      console.error(`Problem with request: ${e.message}`);
    });

    fileStream.pipe(request);
    fileStream.on("end", () => {
      request.end();
    });
  } catch (error) {
    console.error("Error:", error);
  }
})();
